### DISCUSSION TITLE
SwiftUI FocusState

### DISCUSSION THREAD

**Discussion by zhenyataran** [pointfreeco/swift-composable-architecture#2516] [Q&A] [answered]
Can someone describe please how to use this function with simple TextField and Bool property for Focusing
  `
  extension View {
  func synchronize<Root, Value>(
    _ first: Root,
    _ firstKeyPath: ReferenceWritableKeyPath<Root, Value>,
    _ second: Root,
    _ secondKeyPath: ReferenceWritableKeyPath<Root, Value>
  ) -> some View
  where Value: Equatable { … }
  `
  
  Example: 
  `struct PhoneNumberReducer: Reducer {
    
    struct State: Equatable {
        @BindingState var text: String = ""
        var isFocused: Bool = false
    }
    
    enum Action: Equatable, BindableAction {
        case phoneNumberChanged(String)
        case binding(BindingAction<State>)
    }
    
    var body: some Reducer<State, Action> {
        BindingReducer()
        Reduce { state, action in
            switch action {
            case .phoneNumberChanged(let string):
                state.text = string.formatTextIntoPhoneNumber()
                return .none
            case .binding:
                return .none
            }
        }
    }
}
`

`struct PhoneNumberView: View {
    @FocusState var focusedField: Bool
    let store: StoreOf<PhoneNumberReducer>

    var body: some View {
        WithViewStore(self.store, observe: { $0 }) { viewStore in
            TextField("foobar",
                      text: viewStore.$text,
                      prompt: Text("foobar")
                .font(Fonts.Gilroy.semiBold.swiftUIFont(size: 16))
                .foregroundColor(Resources.Colors.textDark.swiftUIColor)
            )
            .onChange(of: viewStore.$text) {
                viewStore.send(.phoneNumberChanged($0.wrappedValue))
            }
            .keyboardType(.phonePad)
            .padding(.bottom, 2)
        }
    }
}`

**Comment by mbrandonw** [comment, marked as answer]
Hi @zhenyataran, we have a [case study](https://github.com/pointfreeco/swift-composable-architecture/blob/20089ee985b04b1ae82e9742aa9d9c8f044700c5/Examples/CaseStudies/SwiftUICaseStudies/01-GettingStarted-FocusState.swift) that demonstrates how to use `@FocusState` in a TCA feature. Essentially you will hold the focus in your feature's `State` struct, _and_ you need to hold `@FocusState` in the view, and then you synchronize the two via the [`bind` view modifier](https://github.com/pointfreeco/swift-composable-architecture/blob/20089ee985b04b1ae82e9742aa9d9c8f044700c5/Examples/CaseStudies/SwiftUICaseStudies/01-GettingStarted-FocusState.swift#L71).

**Reply by zhenyataran** [reply]
Great! Thanks, that's what I needed.
Maybe it will be good to upgrade articles, because when you try to google it, the first link will be this: [SwiftUI Focus State](https://www.pointfree.co/episodes/ep155-swiftui-focus-state).
